File: Engine/Axis/Source/Engine/Entities/Mechanisms/Geometry/AxHeightmapGeometryMechanism.cpp
//---------------------------------------------------------------------------------------------
//	Axis 3D Engine Copyright (C) 2017 Palitri
//
//	This software is released under the MIT License http://palitri.com/axis/license/
//
//	For help and documentation, visit http://palitri.com
//---------------------------------------------------------------------------------------------

#include "AxHeightmapGeometryMechanism.h"

#include "..\..\Textures\AxTexture2D.h"
#include "..\..\Mesh\AxMesh.h"

#include "..\..\..\Primitive\AxVector4.h"

#include "..\..\..\Utilities\Geometry\AxPlaneGeometry.h"
#include "..\..\..\Utilities\Geometry\AxHeightMapGeometry.h"

#include "..\..\..\Utilities\TexCoords\AxPlanarTexCoords.h"

#include "..\..\..\Utilities\Normals\AxNormalsGenerator.h"


AxHeightmapGeometryMechanism::AxHeightmapGeometryMechanism(void)
	: AxPropertyChangeNotifiedMechanism()
{
	this->typeId = AxHeightmapGeometryMechanism::classId;

    this->properties.Add(new AxProperty("Width", 100.0f));
    this->properties.Add(new AxProperty("Length", 100.0f));
    this->properties.Add(new AxProperty("Height", 25.0f));
    this->properties.Add(new AxProperty("Width segments", 100));
    this->properties.Add(new AxProperty("Length segments", 100));
    this->properties.Add(new AxProperty("Color filter", AxVector4(1.0f, 1.0f, 1.0f, 0.0f)));
	this->properties.Add(new AxProperty("Height map", 0, AxParameterType_ReferenceTexture));
    this->properties.Add(new AxProperty("Mesh", 0, AxParameterType_ReferenceMesh));
}


AxHeightmapGeometryMechanism::~AxHeightmapGeometryMechanism(void)
{
}

void AxHeightmapGeometryMechanism::OnPropertiesChanged()
{
	float width = this->properties[AxHeightmapGeometryMechanism::propertyIndex_Width]->GetFloat();
	float length = this->properties[AxHeightmapGeometryMechanism::propertyIndex_Length]->GetFloat();
	float height = this->properties[AxHeightmapGeometryMechanism::propertyIndex_Height]->GetFloat();
	int widthSegments = this->properties[AxHeightmapGeometryMechanism::propertyIndex_WidthSegments]->GetInt();
	int lengthSegments = this->properties[AxHeightmapGeometryMechanism::propertyIndex_LengthSegments]->GetInt();
	AxVector4 colorFilter = this->properties[AxHeightmapGeometryMechanism::propertyIndex_ColorFilter]->GetColor();
	AxTexture2D *heightMap = (AxTexture2D*)this->properties[AxHeightmapGeometryMechanism::propertyIndex_HeightMap]->GetEffectiveValue();
	AxMesh *mesh = (AxMesh*)this->properties[AxHeightmapGeometryMechanism::propertyIndex_Mesh]->GetEffectiveValue();

	if ((heightMap == 0) || (mesh == 0))
		return;

    // Create a flat plane
    AxPlaneGeometry(*mesh->deviceMesh, widthSegments, lengthSegments, AxVector3(width, 0, 0), AxVector3(0, 0, length));
    AxPlanarTexCoords(*mesh->deviceMesh, AxVector3(width, 0, 0), AxVector3(0, 0, length));
    AxNormalsGenerator(*mesh->deviceMesh);

    // Heighten the flat plane along its normals
    AxImage *heightMapImage = heightMap->ToImage();
    AxHeightMapGeometry(*mesh->deviceMesh, *heightMapImage, colorFilter, height);
    delete heightMapImage;

    mesh->ComputeBounds();
    mesh->ComputeNormals();
    mesh->ComputeTangents();

    mesh->deviceMesh->UpdateVertices(0, mesh->deviceMesh->GetVertexCount());
    mesh->deviceMesh->UpdateIndices(0, mesh->deviceMesh->GetIndexCount());

	mesh->serializeVertices = false;
	mesh->serializeNormals = false;
	mesh->serializeTexCoords = false;
	mesh->serializeTangents = false;
	mesh->serializeBones = false;
}

File: Web/Engine/Entities/Mechanisms/Geometry/AxHeightmapGeometryMechanism.js
/**
 * Creates a new heightmap geometry mechanism
 */
function AxHeightmapGeometryMechanism()
{
    AxPropertyChangeNotifiedMechanism.call(this);

    this.typeId = AxHeightmapGeometryMechanism.classId;

    this.properties.Add(new AxProperty(new AxString("Width"), 100.0));
    this.properties.Add(new AxProperty(new AxString("Length"), 100.0));
    this.properties.Add(new AxProperty(new AxString("Height"), 25.0));
    this.properties.Add(new AxProperty(new AxString("Width segments"), 100));
    this.properties.Add(new AxProperty(new AxString("Length segments"), 100));
    this.properties.Add(new AxProperty(new AxString("Color filter"), new AxVector4(1.0, 1.0, 1.0, 0.0)));
    this.properties.Add(new AxProperty(new AxString("Height map"), 0, AxParameterType.ReferenceTexture));
    this.properties.Add(new AxProperty(new AxString("Mesh"), 0, AxParameterType.ReferenceMesh));
}

AxHeightmapGeometryMechanism.prototype = Object.create(AxPropertyChangeNotifiedMechanism.prototype);


AxHeightmapGeometryMechanism.classId = AxMechanism.classId | 206;

AxHeightmapGeometryMechanism.propertyIndex_Width		= AxResource.propertyIndex_ChildPropertiesIndex + 0;
AxHeightmapGeometryMechanism.propertyIndex_Length		= AxResource.propertyIndex_ChildPropertiesIndex + 1;
AxHeightmapGeometryMechanism.propertyIndex_Height		= AxResource.propertyIndex_ChildPropertiesIndex + 2;
AxHeightmapGeometryMechanism.propertyIndex_WidthSegments	= AxResource.propertyIndex_ChildPropertiesIndex + 3;
AxHeightmapGeometryMechanism.propertyIndex_LengthSegments	= AxResource.propertyIndex_ChildPropertiesIndex + 4;
AxHeightmapGeometryMechanism.propertyIndex_ColorFilter		= AxResource.propertyIndex_ChildPropertiesIndex + 5;
AxHeightmapGeometryMechanism.propertyIndex_HeightMap		= AxResource.propertyIndex_ChildPropertiesIndex + 6;
AxHeightmapGeometryMechanism.propertyIndex_Mesh			= AxResource.propertyIndex_ChildPropertiesIndex + 7;


AxHeightmapGeometryMechanism.prototype.OnPropertiesChanged = function()
{
    var width = this.properties.Get(AxHeightmapGeometryMechanism.propertyIndex_Width).GetFloat();
    var length = this.properties.Get(AxHeightmapGeometryMechanism.propertyIndex_Length).GetFloat();
    var height = this.properties.Get(AxHeightmapGeometryMechanism.propertyIndex_Height).GetFloat();
    var widthSegments = this.properties.Get(AxHeightmapGeometryMechanism.propertyIndex_WidthSegments).GetInt();
    var lengthSegments = this.properties.Get(AxHeightmapGeometryMechanism.propertyIndex_LengthSegments).GetInt();
    var colorFilter = this.properties.Get(AxHeightmapGeometryMechanism.propertyIndex_ColorFilter).GetColor();
    var heightMap = this.properties.Get(AxHeightmapGeometryMechanism.propertyIndex_HeightMap).GetEffectiveValue();
    var mesh = this.properties.Get(AxHeightmapGeometryMechanism.propertyIndex_Mesh).GetEffectiveValue();
    

    if ((heightMap === 0) || (mesh === 0))
        return;

    // Tweak for asynchrnously loaded images, which might not yet be loaded in the first few frames
    if ((heightMap.width === 0) || (heightMap.height === 0))
    {
        var oldParam = this.oldProperties.Get(AxHeightmapGeometryMechanism.propertyIndex_HeightMap);
        oldParam.SetValue(0, oldParam.type);
        return;
    }

    // Create a flat plane
    new AxPlaneGeometry(mesh.deviceMesh, widthSegments, lengthSegments, new AxVector3(width, 0, 0), new AxVector3(0, 0, length));
    new AxPlanarTexCoords(mesh.deviceMesh, new AxVector3(width, 0, 0), new AxVector3(0, 0, length));
    new AxNormalsGenerator(mesh.deviceMesh);

    // Heighten the flat plane along its normals
    var heightMapImage = heightMap.ToImage();
    new AxHeightMapGeometry(mesh.deviceMesh, heightMapImage, colorFilter, height);

    mesh.ComputeBounds();
    mesh.ComputeNormals();
    mesh.ComputeTangents();

    mesh.deviceMesh.UpdateVertices(0, mesh.deviceMesh.GetVertexCount());
    mesh.deviceMesh.UpdateIndices(0, mesh.deviceMesh.GetIndexCount());

    mesh.serializeVertices = false;
    mesh.serializeNormals = false;
    mesh.serializeTexCoords = false;
    mesh.serializeTangents = false;
    mesh.serializeBones = false;
};
